core/tauri/src/ipc/protocol.rs

```diff
@@ -0,0 +1,272 @@
+      options: Option<RequestOptions>,
+    }
+
+    #[allow(unused_mut)]
+    let mut invoke_message: Option<crate::Result<Message>> = None;
+
+    #[cfg(feature = "isolation")]
+    {
+      #[derive(serde::Deserialize)]
+      struct IsolationMessage<'a> {
+        cmd: String,
+        callback: CallbackFn,
+        error: CallbackFn,
+        #[serde(flatten)]
+        payload: crate::utils::pattern::isolation::RawIsolationPayload<'a>,
+        options: Option<RequestOptions>,
+      }
+
+      if let crate::Pattern::Isolation { crypto_keys, .. } = manager.pattern() {
+        invoke_message.replace(
+          serde_json::from_str::<IsolationMessage<'_>>(&message)
+            .map_err(Into::into)
+            .and_then(|message| {
+              Ok(Message {
+                cmd: message.cmd,
+                callback: message.callback,
+                error: message.error,
+                payload: serde_json::from_slice(&crypto_keys.decrypt(message.payload)?)?,
+                options: message.options,
+              })
+            }),
+        );
+      }
+    }
+
+    match invoke_message
+      .unwrap_or_else(|| serde_json::from_str::<Message>(&message).map_err(Into::into))
+    {
+      Ok(message) => {
+        let _ = window.on_message(InvokeRequest {
+          cmd: message.cmd,
+          callback: message.callback,
+          error: message.error,
+          body: message.payload.into(),
+          headers: message.options.map(|o| o.headers.0).unwrap_or_default(),
+        });
+      }
+      Err(e) => {
+        let _ = window.eval(&format!(
+          r#"console.error({})"#,
+          serde_json::Value::String(e.to_string())
+        ));
+      }
+    }
+  }
+}
+
+fn handle_ipc_request<R: Runtime>(
+  request: &HttpRequest,
+  manager: &WindowManager<R>,
+  label: &str,
+) -> std::result::Result<InvokeResponse, String> {
+  if let Some(window) = manager.get_window(label) {
+    // TODO: consume instead
+    #[allow(unused_mut)]
+    let mut body = request.body().clone();
+
+    let cmd = request
+      .uri()
+      .strip_prefix("ipc://localhost/")
+      .map(|c| c.to_string())
+      // the `strip_prefix` only returns None when a request is made to `https://tauri.$P` on Windows
+      // where `$P` is not `localhost/*`
+      // in this case the IPC call is considered invalid
+      .unwrap_or_else(|| "".to_string());
+    let cmd = percent_encoding::percent_decode(cmd.as_bytes())
+      .decode_utf8_lossy()
+      .to_string();
+
+    // the body is not set if ipc_custom_protocol is not enabled so we'll just ignore it
+    #[cfg(all(feature = "isolation", ipc_custom_protocol))]
+    if let crate::Pattern::Isolation { crypto_keys, .. } = manager.pattern() {
+      match crate::utils::pattern::isolation::RawIsolationPayload::try_from(&body)
+        .and_then(|raw| crypto_keys.decrypt(raw))
+      {
+        Ok(data) => body = data,
+        Err(e) => {
+          return Err(e.to_string());
+        }
+      }
```

Comment:
not a security concern, but this could be written a little cleaner IMO
```suggestion
      body = crate::utils::pattern::isolation::RawIsolationPayload::try_from(&body)
        .and_then(|raw| crypto_keys.decrypt(raw))
        .map_err(|e| e.to_string())?;
```